### PR TITLE
Receive: Tenants' external labels proposal

### PR DESCRIPTION
**Pull request by haanhvu**
Signed-off-by: haanhvu <haanh6594@gmail.com>

Proposal for #5434

### CODE REVIEW

**Comment by haanhvu** [comment]
@fpetkovski @saswatamcode This probably needs more details. Let's discuss this tomorrow.

File: docs/proposals-accepted/202209-receive-tenant-external-labels.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+## 2 What problem(s) are we trying to solve?
+
+We want to add arbitrary external labels to each tenant in Thanos Receiver.
+
+## 3 How is it going to be helpful?
+
+This new functionality is helpful in many use cases. Two typical ones are:
+* Compact tenant blocks based on tenants’ external labels
+* Query data from tenants which share the same external labels
```
Comment (yeya24) [inline docs/proposals-accepted/202209-receive-tenant-external-labels.md:21]: Does the existing `tenant_id` work in this case? I understand the second query usecase as users may want to have some cross-tenants aggregation queries or query tenants with some other labels like `region`.

For the first usecase, when do we want to compact based on other external labels? I think we should always compact using `tenant_id`, right?

File: docs/proposals-accepted/202209-receive-tenant-external-labels.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+## 4 Pitfalls of the current solution
+
+Currently, we can only add external labels to Receivers. Each tenant in a Receiver only gets one external label, which is the tenant ID. This is not flexible enough to do complex compact and query works at the tenant level.
+
```
Comment (saswatamcode) [inline docs/proposals-accepted/202209-receive-tenant-external-labels.md:26]: Maybe we can add a Goals/Non-goals here?

File: docs/proposals-accepted/202209-receive-tenant-external-labels.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+## 3 How is it going to be helpful?
+
+This new functionality is helpful in many use cases. Two typical ones are:
+* Compact tenant blocks based on tenants’ external labels
+* Query data from tenants which share the same external labels
+
+## 4 Pitfalls of the current solution
+
+Currently, we can only add external labels to Receivers. Each tenant in a Receiver only gets one external label, which is the tenant ID. This is not flexible enough to do complex compact and query works at the tenant level.
```
Comment (saswatamcode) [inline docs/proposals-accepted/202209-receive-tenant-external-labels.md:25]: We cannot partition tenant tsdb blocks yet, but this change would enable that.

File: docs/proposals-accepted/202209-receive-tenant-external-labels.md [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+## 5 What are the possible solution approaches?
```
Comment (saswatamcode) [inline docs/proposals-accepted/202209-receive-tenant-external-labels.md:27]: Subsection "How"?

File: docs/proposals-accepted/202209-receive-tenant-external-labels.md [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+We’re seeing two solution approaches:
+* Specify the tenants for each external label in Receiver: Since we already have the functionality of adding external labels to Receiver, is there a way to specify which tenants each external label goes to? Because many tenants may share the same external labels, this may be a straightforward way for users.
```
Comment (saswatamcode) [inline docs/proposals-accepted/202209-receive-tenant-external-labels.md:30]: Maybe if you have questions, you can organize them in an open questions section below?
Right now it feels like this is actual solution to the problem, and an "answer" rather than a question 🙂

File: docs/proposals-accepted/202209-receive-tenant-external-labels.md [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+* Add external labels to each tenant directly: This way, we treat each tenant independently, which seems to be a more straightforward way in implementation. If goinf this way, we can start with the tenant configuration file.
```
Comment (saswatamcode) [inline docs/proposals-accepted/202209-receive-tenant-external-labels.md:31]:
```suggestion
* Add external labels to each tenant directly: This way, we treat each tenant independently, which seems to be a more straightforward way in implementation. If going this way, we can start with the tenant configuration file.
```
Maybe linking documentation regarding tenant configuration file and showing how you want to add a new field to this file might be better? This would give reader a much better idea of the changes that you're proposing.

File: docs/proposals-accepted/202209-receive-tenant-external-labels.md [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+## 6 Which solution approach is the most favorable?
+
+After discussing internally in the issue team (@fpetkovski, @saswatamcode, @haanhvu), we agree the second approach may be the better one to start with since it’s more straightforward in implementation.
+
+## 7 Potential challenges of the favorable approach?
+
+We’re seeing two main challenges:
+* Support external labels for decoupled routers and ingesters: Receive routers and ingestors are decoupled functionally, but currently they are not decoupled in code. Issue [#5643](https://github.com/thanos-io/thanos/issues/5643) has already been raised for this. With routers and ingestors intermingled in code, we need to find a way to support external labels for RouterOnly and IngestorOnly. One apparent way is to fix [#5643](https://github.com/thanos-io/thanos/issues/5643) at least partially enough before tackling this challenge.
```
Comment (saswatamcode) [inline docs/proposals-accepted/202209-receive-tenant-external-labels.md:40]: Maybe this could be distributed within "Goals" or "Pitfalls"?

**Review by saswatamcode — COMMENTED** [review]
Thanks for awesome work!

Some comments!

**Comment by haanhvu** [comment]
@fpetkovski I'll push something here at the end of the week. So don't review this yet if you don't have much time.

File: docs/proposals-accepted/202209-receive-tenant-external-labels.md (hunk 5 above)
Comment (haanhvu) [reply docs/proposals-accepted/202209-receive-tenant-external-labels.md:30]: Yeah that's not a question anymore. I answered it myself after reading the code :D

File: docs/proposals-accepted/202209-receive-tenant-external-labels.md (hunk 3 above)
Comment (haanhvu) [reply docs/proposals-accepted/202209-receive-tenant-external-labels.md:25]: @fpetkovski If we don't persist tenant's external labels in TSDB, can we still partition tenant TSDB blocks?

File: docs/proposals-accepted/202209-receive-tenant-external-labels.md (hunk 1 above)
Comment (haanhvu) [reply docs/proposals-accepted/202209-receive-tenant-external-labels.md:21]: If you do some query based on some other external label, and you know that you'll do that query a lot, would compact based on that external label make your query faster? That's my understanding.

Comment (yeya24) [reply docs/proposals-accepted/202209-receive-tenant-external-labels.md:21]: Compact blocks means merging blocks together to a single block. Each tenant's blocks should only merge with their own blocks. I don't think it is correct to merge blocks across tenants.
I think what you want is to have some kind of secondary index to group some tenants blocks together? That's just logical grouping, right? Not compaction.

File: docs/proposals-accepted/202209-receive-tenant-external-labels.md (hunk 3 above)
Comment (yeya24) [reply docs/proposals-accepted/202209-receive-tenant-external-labels.md:25]: I am a little bit confused. The partition is to partition each TSDB block?

File: docs/proposals-accepted/202209-receive-tenant-external-labels.md (hunk 1 above)
Comment (fpetkovski) [reply docs/proposals-accepted/202209-receive-tenant-external-labels.md:21]: The querier can filter out stores which do not match certain external labels, so maybe we need to rephrase this to say:
* Reduce fanout when querying data from tenants which share certain external labels

**Comment by haanhvu** [comment]
@fpetkovski @saswatamcode @yeya24 I made some changes. Please help review.

File: docs/proposals-accepted/202209-receive-tenant-external-labels.md [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+## 8 Implementation plan
+
+* Add a new `external_labels` field in the hasring config
+* Map tenant ID to tenant’s external labels
+* Extend tenant’s external labels to the list of labels when the tenant’s TSDB instance is created
+* Implement tenants’ external labels in RouterIngestor
+* Handle changes for tenant’s external labels
+* Implement tenants’ external labels in RouterOnly
+*Implement tenants’ external labels in IngestorOnly
```
Comment (yeya24) [inline docs/proposals-accepted/202209-receive-tenant-external-labels.md:65]: Nit, we need one space here

File: docs/proposals-accepted/202209-receive-tenant-external-labels.md [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+        "endpoints": ["tenant-a-1.metrics.local:19291/api/v1/receive", "tenant-a-2.metrics.local:19291/api/v1/receive"],
+        "tenants": ["tenant-a"]
+        "external_labels": ["label-1", "label-2", "label-3"]
+    },
+ ]
+ ```
+In [MultiTSDB](https://github.com/thanos-io/thanos/blob/4ce3fe19ebb39a308769fb2a9492295b1f113701/pkg/receive/multitsdb.go#L46), each tenant’s external labels will be mapped to its tenant ID.
+Each tenant’s external labels will also be extended to the [list of labels](https://github.com/thanos-io/thanos/blob/4ce3fe19ebb39a308769fb2a9492295b1f113701/pkg/store/labelpb/label.go#L282) when the tenant’s TSDB instance is created.
+Tenants’ external labels will be first implemented in RouterIngestor, since this is the main mode.
+Once tenants’ external labels are implemented in RouterIngestor, we’ll handle changes in tenants’ external labels. So that when users change the values of `external_labels` field in the hasring config, the changes will appear on users’ query.
```
Comment (fpetkovski) [inline docs/proposals-accepted/202209-receive-tenant-external-labels.md:53]:
```suggestion
Once tenants’ external labels are implemented in RouterIngestor, we will handle changes in tenants’ external labels. So that when users change the values of `external_labels` field in the hashring config, the changes will appear on users’ query.
```

**Review by fpetkovski — COMMENTED** [review]
I like this proposal, simple and straight to the point. We need to polish out some typos, but other than that looks quite good already

File: docs/proposals-accepted/202209-receive-tenant-external-labels.md [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+* Backward compatibility (e.g., with Receiver’s external labels) is assured
+* Tenants’ external labels are handled separately in RouterIngestor, RouterOnly, and IngestorOnly modes
+
+## 5 Non-goals
+
+* Logically split RouterOnly and IngestorOnly modes (Issue [#5643](https://github.com/thanos-io/thanos/issues/5643)): If RouterOnly and IngestorOnly modes are logically split, implementing tenants’ external labels in RouterOnly and IngestorOnly modes would be less challenging. However, the split could be tricky and time-consuming, which might distract us from implementing tenants’ external labels. So we will not do it now.
+
+## 6 Audience
+
+Users who need to have cross-tenant information and do cross-tenant activities
```
Comment (saswatamcode) [inline docs/proposals-accepted/202209-receive-tenant-external-labels.md:35]: This isn't really cross-tenancy. Maybe something like this would be better?
```suggestion
Users who are admin personas and need to perform admin operations on Thanos for multiple tenants
```

File: docs/proposals-accepted/202209-receive-tenant-external-labels.md [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+
+* Allow users to add arbitrary tenants’ external labels in the easiest way possible
+* Allow users to change arbitrary tenants’ external labels in the easiest way possible
+* Changes in tenants’ external labels are handled correctly
+* Backward compatibility (e.g., with Receiver’s external labels) is assured
+* Tenants’ external labels are handled separately in RouterIngestor, RouterOnly, and IngestorOnly modes
+
+## 5 Non-goals
+
+* Logically split RouterOnly and IngestorOnly modes (Issue [#5643](https://github.com/thanos-io/thanos/issues/5643)): If RouterOnly and IngestorOnly modes are logically split, implementing tenants’ external labels in RouterOnly and IngestorOnly modes would be less challenging. However, the split could be tricky and time-consuming, which might distract us from implementing tenants’ external labels. So we will not do it now.
```
Comment (saswatamcode) [inline docs/proposals-accepted/202209-receive-tenant-external-labels.md:31]:
```suggestion
* Logically split RouterOnly and IngestorOnly modes (Issue [#5643](https://github.com/thanos-io/thanos/issues/5643)): If RouterOnly and IngestorOnly modes are logically split, implementing tenants’ external labels in RouterOnly and IngestorOnly modes would be less challenging. However, this needs more in-depth investigation and will be tackled as a separate issue.
```

**Review by saswatamcode — COMMENTED** [review]
Thanks! Great work! 🙂 

Some small nits and typos from above.

File: docs/proposals-accepted/202209-receive-tenant-external-labels.md [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+## 7 How
+
+We’ll add a new `external_labels` field in the hasring config. Something like this:
+ ```
+ [
+    {
+        "hashring": "tenant-a",
+        "endpoints": ["tenant-a-1.metrics.local:19291/api/v1/receive", "tenant-a-2.metrics.local:19291/api/v1/receive"],
+        "tenants": ["tenant-a"]
+        "external_labels": ["label-1", "label-2", "label-3"]
```
Comment (haanhvu) [inline docs/proposals-accepted/202209-receive-tenant-external-labels.md:46]: @fpetkovski @saswatamcode here should it be `"external_labels": ["label-name-1:label-value-1", "label-name-2:label-value-2", "label-name-3:label-value-3"]`? 🤔

Comment (saswatamcode) [reply docs/proposals-accepted/202209-receive-tenant-external-labels.md:46]: I think it can be similar to how we pass in label key-value pairs via flag like `--label='receive_replica="0"'`.
So it would be something like `external_labels": ["label=value", "label2=value2"]`.

File: docs/proposals-accepted/202209-receive-tenant-external-labels.md (hunk 3 above)
Comment (haanhvu) [reply docs/proposals-accepted/202209-receive-tenant-external-labels.md]: The word "partition" might lead us to thinking of storage partitioning. So I decided to use "select" instead.

File: docs/proposals-accepted/202209-receive-tenant-external-labels.md (hunk 12 above)
Comment (saswatamcode) [reply docs/proposals-accepted/202209-receive-tenant-external-labels.md:46]: Yes, so I think you can reuse `parseFlagLabels` https://github.com/thanos-io/thanos/blob/main/cmd/thanos/rule.go#L747 to basically parse string slice from YAML.

Comment (haanhvu) [reply docs/proposals-accepted/202209-receive-tenant-external-labels.md:46]: @saswatamcode Is the quote surrounding the label value important?

Comment (yeya24) [reply docs/proposals-accepted/202209-receive-tenant-external-labels.md:46]: So `"external_labels": ["label-1", "label-2", "label-3"]` is specifying the label name value pairs? or just label names?

**Review by yeya24 — COMMENTED** [review]
I am wondering if the following dynamic external label is part of the scope or not. 
User can specify the external label names they want to extract as external labels to blocks. Upon receiving series from remote write we automatically group new blocks based on the external label value.

This sounds kind of hard but would be a super useful feature not only in Thanos but also in Cortex.

**Comment by fpetkovski** [comment]
We can start with statically defined labels first. Once that is in place, it would be interesting to explore extracting external labels from timeseries data.

**Comment by yeya24** [comment]
Hey @haanhvu, could you please address the comments and we can merge the proposal?

**Comment by haanhvu** [comment]
> Hey @haanhvu, could you please address the comments and we can merge the proposal?

@yeya24 Yes. I'll also add your idea as a follow-up^^

**Comment by yeya24** [comment]
Ping @haanhvu. It is better to resolve the comments and merge the proposal before merging the implementation pr.

**Comment by haanhvu** [comment]
@yeya24 Yeah sorry for the delay I'm taking some days off. I update this proposal this week don't worry^^

**Comment by haanhvu** [comment]
@fpetkovski @saswatamcode @yeya24 I updated the proposal. Besides the suggested wording changes, I also made some other changes based on what we've learned and discussed in the past months.

Please help review!

**Review by yeya24 — APPROVED** [review]
Thanks @haanhvu. This looks good to me. Anyone wants to take another review before I merge it?

**Review by matej-g — APPROVED** [review]

**Review by yeya24 — APPROVED** [review]
Thanks. Let's fix doc on main.
